### PR TITLE
First step cleaning up the merchant data.

### PR DESCRIPTION
**Pull request by boris-tschirschwitz**

### CODE REVIEW

**Comment by boris-tschirschwitz** [comment]
Of course the transformations I am doing on the merchant data are all up for discussion, it's just a suggestion.

File: clean_merchants.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    # There are a lot (over 100000) merchants with incomplete city and state data.
+    # Since it seems unlikely that the location has an influence on the loyalty, we drop these
+    # Also, category_2 is always NaN when the city is unknown, so we drop that, too.
+    merchants_df.drop(columns=['city_id', 'state_id', 'category_2'], inplace=True)
+
+    # There are only 15 merchants where some of the lag data is invalid, so let's just drop these.
+    merchants_df = merchants_df[np.isnan(merchants_df['avg_sales_lag3']) == False]
+
+    # Category_1 and _4 have only values 'N' and 'Y'. Let's replace this with Booleans False and True
```
Comment (erwald) [inline clean_merchants.py]: You can also set the fields to be categorical ones (as seen by pandas) in this way:

```python
for v in cat_vars:
    df[v] = df[v].astype('category').cat.as_ordered()
```

File: clean_merchants.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    merchants_df['category_1'] = merchants_df['category_1'].map({'Y': True, 'N': False})
+    merchants_df['category_4'] = merchants_df['category_4'].map({'Y': True, 'N': False})
+
+    # The 'most recent' columns measure the size of transactions from 'A' the most to 'E" the least.
+    # Let's translate this into 4.0 to 0.0 to make the range continuous.
+    range_to_num = {'A': 4, 'B': 3, 'C': 2, 'D': 1, 'E': 0}
```
Comment (erwald) [inline clean_merchants.py]: 👍

File: clean_merchants.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    merchants_df['most_recent_sales_range'] = merchants_df['most_recent_sales_range'].map(range_to_num)
+
+    # We need to somehow unify the data for one merchant with several entries.
+    # The easiest case is where the all the identifying data is equal.
+    # Here we simply aggregate over the features:
+    # - numerical_1 and _2: These seem to be measuring some kind of continuous value, let's take the mean.
+    # - category_1: This is always True or False, but only 7033 are True. So let's keep True as more special.
```
Comment (erwald) [inline clean_merchants.py]: Or we can introduce a third category, giving us `True`, `False` and `Both` as possible values.

On a related note, I thought I would quickly go through how embedding of categorical columns works internally in the fastai library on Thursday.

File: clean_merchants.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    #               max(True, False) = True
+    # - most_recent_sales_range,
+    #   most_recent_purchases_range: More sales/purchases seems to translate to higher ranges,
+    #                                so let's use the maximum b/c we're adding transactions.
+    # - average_X_lagY: This one is tricky. Since the average is divided by the value of the last
+    #                   active month, it measures how th last active month was worse than the average.
```
Comment (erwald) [inline clean_merchants.py]: th -> the

File: clean_merchants.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    #                   Maybe we can be more precise by actually analysing the transactions,
+    #                   for now I'll take the mean.
+    # - active_months_lagX: Take the maximum b/c we're adding transactions.
+    # - category_4: False : True is more than 2:1, so let's take True over False
```
Comment (erwald) [inline clean_merchants.py]: Same thing, we could perhaps live with having three categorical options here.

File: clean_merchants.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    grouped = merchants_df.groupby(['merchant_id', 'merchant_group_id', 'merchant_category_id', 'subsector_id'])
+    aggregated_df = grouped.agg({'numerical_1': np.mean,
```
Comment (erwald) [inline clean_merchants.py]: 👍 This is the kinda stuff that makes it easy to be productive with pandas, once you know all of the arcane incantations it provides.

File: clean_merchants.py (hunk 1 above)
Comment (boris-tschirschwitz) [reply clean_merchants.py]: Interesting. That seems especially useful for categories with 3 and more values,

File: clean_merchants.py (hunk 3 above)
Comment (boris-tschirschwitz) [reply clean_merchants.py]: That might be a good option.
👍  for going over embeddings.

File: clean_merchants.py (hunk 6 above)
Comment (boris-tschirschwitz) [reply clean_merchants.py]: I think I'd prefer to simply define a mapping from one group of rows to an aggregated new row. Maybe apply can do that on groupby objects, but I didn't study that too much yet,

File: clean_merchants.py (hunk 5 above)
Comment (boris-tschirschwitz) [reply clean_merchants.py]: I liked that I could go with build in aggregation functions :-)
I'll have a look at it.

File: clean_merchants.py (hunk 1 above)
Comment (boris-tschirschwitz) [reply clean_merchants.py]: I now changed both categories to categorical data with these three values.
It turns out, for my first step I didn't really need it as all groups take only one of 'Y' or 'N'.

But I am pretty sure that changes when I'm trying to further unify on merchant_id only.

File: clean_merchants.py (hunk 3 above)
Comment (boris-tschirschwitz) [reply clean_merchants.py]: Done, see above.

**Comment by boris-tschirschwitz** [comment]
I'll probably also want to change the 'A' to 'E' columns to categorical data, in this case ordered.

**Comment by erwald** [comment]
I was thinking, by the way, that a good start to aggregating all these categorical variables is to simply take the most frequently occurring one. Say, for card `123`, taking the city id (or whatever) of the store where the owner of that card did most of their purchases with it. [Here's how we can do that](https://stackoverflow.com/q/15222754/1077679), for future reference.

**Comment by boris-tschirschwitz** [comment]
> I was thinking, by the way, that a good start to aggregating all these categorical variables is to simply take the most frequently occurring one. Say, for card `123`, taking the city id (or whatever) of the store where the owner of that card did most of their purchases with it. [Here's how we can do that](https://stackoverflow.com/q/15222754/1077679), for future reference.

Damn, I totally missed that we also have the merchant category and location data in the transactions (that's why I didn't link them in the realtime board). let's see how well they match.

**Comment by boris-tschirschwitz** [comment]
I have looked at matching the columns ['merchant_id', 'merchant_category_id', 'subsector_id', 'city_id', 'state_id', 'category_2'] of merchants and historical transactions, but whenever I take more columns than just 'merchant_id' I get more unique rows in the transactions than in the merchants. Therefore only if I match on 'merchant_id' alone is there a chance of finding a merchant for every transaction.

**Comment by boris-tschirschwitz** [comment]
Also, there are no merchant entries where we can deduce a missing city ID from the state ID and another entry with city and state.

**Comment by boris-tschirschwitz** [comment]
Even if we just look at city/state pairs independent of the merchants in the merchant data there are no city IDs that have one entry with state ID and some without, such that we cold fill in the missing state IDs.

**Comment by erwald** [comment]
So I think it would make more sense to think of the state and city id fields more as categories than anything that will actually help us identify a merchant, transaction or anything else, or is that a misconception?

**Pull request merged**